### PR TITLE
feat: Add support for resolving UDPs

### PR DESCRIPTION
**Pull request by GeraldIr**
This PR adds the ability to "resolve" process graphs. Resolving entails replacing any encountered UDPs (User Defined Processes) with their implementations until only predefined processes remain which any given executor can handle.

### CODE REVIEW

**Review by LukeWeidenwalker — CHANGES_REQUESTED** [review]
Here's a first round of review - heads-up that I haven't fully gotten my head around the actual json manipulations, so there'll be more feedback on that later, but wanted to at least partially unblock you on this!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py:92]: We don't have access to the user_id in the resolving logic, but I could look for the process_id across all namespaces I guess.

I just figured that having it directly accessible would be preferable, but I can change it.

File: openeo_pg_parser_networkx/graph.py
Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/graph.py:447]: Hmm - maybe what we should do is not make `resolve_process_graph` a method of `OpenEOProcessGraph`, but just expose it through the `resolving_utils.py` file. It doesn't really operate on the networkx object anyways, so maybe it's less confusing to just have it as separate functionality!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/resolving_utils.py:92]: Ok, let's look in all namespaces then - thanks!

File: tests/data/res_tests/resolved/resolved_complex.json
Comment (LukeWeidenwalker) [inline tests/data/res_tests/resolved/resolved_complex.json:46]: I don't think that passing parameters in like this is currently supported by https://github.com/Open-EO/openeo-processes-dask/blob/main/openeo_processes_dask/process_implementations/core.py#L16 - would be great if you could double-check this!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (LukeWeidenwalker) [inline openeo_pg_parser_networkx/resolving_utils.py:137]: Lmk what you think, but as far as I understand this, the arguments are only considered from the immediate parent here. According to https://openeo.org/documentation/1.0/developers/api/reference.html#section/Processes/Process-Graphs, ("Accessing process parameters"), parameter references can be resolved from the nearest parent that has it. Is this being addressed in the recursion somehow I'm not getting? Tbh I haven't encountered such a doubly nested process graph yet, but wanted your take on this regardless!

**Review by LukeWeidenwalker — CHANGES_REQUESTED** [review]
Alright, finally finished this - apologies for the delay!
Couple of further comments, but the core resolving logic looks good to me and the tests are good too!

I'd maybe add a TODO comment to the top of the `resolving_utils.py`

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py:92]: Can we assume the uniqueness of the process_id across namespaces? We let users fill up the process_registry themselves after all.

File: tests/data/res_tests/resolved/resolved_complex.json
Comment (GeraldIr) [reply tests/data/res_tests/resolved/resolved_complex.json:46]: This isn't an artifact of resolving. This is just how explicit parameter values are passed by processes with sub-graphs (like apply)

```
{
  "id": "valid_load_small",
  "parameters": [],
  "process_graph": {
    "load7": {
      "process_id": "load_collection",
      "arguments": {
        "id": "boa_sentinel_2",
        "spatial_extent": {
          "west": 16.200432580649483,
          "east": 16.58041390646924,
          "south": 48.04835633006704,
          "north": 48.330435665810796
        },
        "temporal_extent": [
          "2020-07-01T00:00:00Z",
          "2020-07-03T00:00:00Z"
        ],
        "bands": [
          "B02"
        ],
        "properties": {}
      }
    },
    "apply1": {
      "process_id": "apply",
      "arguments": {
        "data": {
          "from_node": "load7"
        },
        "process": {
          "process_graph": {
            "add1": {
              "process_id": "add",
              "arguments": {
                "x": {
                  "from_parameter": "x"
                },
                "y": {
                  "from_parameter": "context"
                }
              },
              "result": true
            }
          }
        },
        "context": 1
      }
    },
    "save2": {
      "process_id": "save_result",
      "arguments": {
        "data": {
          "from_node": "apply1"
        },
        "format": "GTIFF"
      },
      "result": true
    }
  }
}
```
This is a process graph that I threw together quickly in the web-editor (without any UDPs) and the context is in there as well.

I also think it must be supported as I have run the graph (resolved_complex.json) successfully before and if it somehow isn't we should work on supporting it.

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py:137]: I'm actually unsure if there is such a double nesting and how you would achieve it. You can doubly nest UDPs of course, but for those to be valid you'd have to wire the parameters for each level of nesting and then the recursion kicks in when resolving each level. 

So I don't know how a doubly nested UDP with a parameter that only gets used 2 steps down would even be possible to create.

File: openeo_pg_parser_networkx/graph.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/graph.py:447]: Sure, we can do that. That way you also don't have to first load it into a graph object, then resolve it and then use the new graph object, but you can just get the resolved graph object by default.

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/resolving_utils.py:92]: Hmm, don't think we can assume this - actually, is there any rule that prevents users from naming their UDP like a predefined process?

Tbh, as long as we guarantee that predefined processes always get preferred to UDPs if there's a collision, that should be fine.

File: tests/data/res_tests/resolved/resolved_complex.json
Comment (LukeWeidenwalker) [reply tests/data/res_tests/resolved/resolved_complex.json:46]: Right, not related to UDPs, gotcha!
Hmm. I'm not sure tbh, would have to try it out. I'll create an issue to look into it when we have time!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/resolving_utils.py:137]: Right - doesn't seem very useful or likely to appear indeed, so probably not worth spending more time on it - thanks!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py:92]: I meant this more as in multiple UDPs with the same name across different namespaces that get passed into the process_registry out of carelessness. Predefined processes are already preferred over UDPs because if a process is found in the DEFAULT_NAMESPACE it isn't resolved at all (even if it is empty and the user intended for it to be a UDP).

I mean, forcing a 'user' namespace only fixes this problem by forcing users to have a single clear UDP which is guaranteed to be the intended one, but also adds more confusion, because now a special process_registry has to be prepared if this functionality is to be used.

I don't know if these kinds of collisions would ever come up, but maybe we should just disallow backends to pass an already filled process_registry and only allow usage with the callback function. That way we always get a single, clear UDP implementation back and we don't have to think about any naming clashes.

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py:191]: Actually UDPs do have process_id fields, just their implementation does not as it consists of multiple processes and is wrapped in an additional layer, this function functions to detect these additional layers, I guess a better name for it would be "is_process" as that makes its functionality clearer.

Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/resolving_utils.py:191]: Ahh, thanks for resolving my confusion!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py]: I changed this to just using an optional argument "namespace" which defaults to user and gets passed on to the get_udp_spec call.

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py]: Would it be an option to leave this up to whoever is implementing the get_udp_spec to throw an exception inside there? Would have more information in that implementation as to why exactly the UDP was not found.

Comment (LukeWeidenwalker) [reply openeo_pg_parser_networkx/resolving_utils.py]: Ehh, I wouldn't rely on the user to do this nicely, we should still catch this error! You can use exception chaining
e.g. ```except KeyError as exc:
    raise ValueError("Bad grape") from exc```
to raise the original exception with it!

File: pyproject.toml
Comment (LukeWeidenwalker) [inline pyproject.toml:6]: welcome aboard!

File: openeo_pg_parser_networkx/resolving_utils.py
Comment (GeraldIr) [reply openeo_pg_parser_networkx/resolving_utils.py]: The new version should address this, I will hold off on resolving this thread until further comment though.

**Pull request merged**
